server/api/visualization/visualization.controller.js

```diff
@@ -1,73 +1,67 @@
 'use strict';
 
+var _ = require('lodash');
 var AbsenceRecord = require('../absence-record/absence-record.model');
 
-function currentAbsenceRecordPipeline(user) {
-  var pipeline = [];
-  if (user.role === 'teacher') {
-    pipeline.push({
-      $match: {school: user.assignment}
-    });
-  }
-  pipeline.push({
+var categoryDefaults = {
+  'Normal': 0,
+  'ARCA': 0,
+  'CA': 0
+};
+
+var arcaCondition = {
+  $cond: [{
+    $lte: [{$divide: ['$entries.present', '$entries.enrolled']}, 0.9]
+  }, 'ARCA', 'Normal']
+};
+
+var categorizeCondition = {
+  $cond: [{$gte: ['$entries.absences', 20]}, 'CA', arcaCondition]
+};
+
+exports.cfaComparison = function(req, res) {
+  var pipeline = [{
     $sort: {date: -1}
-  });
-  pipeline.push({
-    $group: {
-      _id: '$school',
-      recordId: {$first: '$_id'},
-      date: {$first: '$date'},
-      school: {$first: '$school'},
-      entries: {$first: '$entries'}
-    }
-  });
-  pipeline.push({
+  }, {
+    $group: {_id: '$school', entries: {$first: '$entries'}}
+  }, {
     $unwind: '$entries'
-  });
-  return pipeline;
-}
-
-exports.cfaVsNotcfa = function(req, res) {
-  var pipeline = currentAbsenceRecordPipeline(req.user);
-  pipeline.push({
-    $project: { 
-      _id: 0, 
-      student: '$entries.student', 
-      present: '$entries.present', 
-      enrolled: '$entries.enrolled',
-      absences: '$entries.absences' 
-    }
   }, {
     $project: {
-      student: 1,
-      arca: {
-        $and: [
-          { $lte: [{ $divide: ['$present', '$enrolled'] }, 0.9] },
-          { $lte: ['$absences', 19] }
-        ]
-      },
-      ca: {
-        $gte: ['$absences', 20]
-      }
+      _id: false,
+      student: '$entries.student',
+      category: categorizeCondition
     }
-  });
+  }];
+  if (req.school) {
+    pipeline.unshift({
+      $match: {school: req.school._id}
+    });
+  }
   AbsenceRecord.aggregate(pipeline, function(err, results) {
-    if (err) return handleError(res, err);
     AbsenceRecord.populate(results, {
       path: 'student',
       model: 'Student',
-      select: '_id currentSchool cfa withdrawn active',
-      populate: {
-        path:'currentSchool',
-        model: 'School',
-        select: '-_id name'
-      }
-    }, function(err, result) {
+      select: 'cfa withdrawn',
+      match: {withdrawn: false}
+    }, function(err, docs) {
       if (err) return handleError(res, err);
-      return res.status(200).json(result);
+      res.status(200).json(_(docs)
+        .groupBy('student.cfa')
+        .mapValues(function(students) {
+          return _(students)
+            .groupBy('category')
+            .mapValues('length')
+            .defaults(categoryDefaults)
+            .value();
+        })
+        .mapKeys(function(value, key) {
+          return key === 'true' ? 'cfa' : 'non';
+        })
+        .value());
     });
```

Comment:
The server is refactored to do some reshaping of the data post population because this aggregation should never deal with a lot of documents. The data is shaped in a way that it could be displayed as a table if desired. The client's data visualization controller then reshapes the data into series for use with Highcharts.